projects/continuous_learning/exp_lesparse.py

```diff
@@ -0,0 +1,376 @@
+                use_kwinners_local=use_kwinners_local,
+                consolidated_sparse_weights=csw
+            )
+
+            # Compute next layer input shape
+            wout = (width - 5) + 1
+            maxpool_width = wout // 2
+            current_input_shape = (cnn_out_channels[i], maxpool_width, maxpool_width)
+
+        # Flatten CNN output before passing to linear layer
+        self.add_module("flatten", Flatten())
+
+        # Add Linear layers
+        input_size = np.prod(current_input_shape)
+        for i in range(len(linear_n)):
+            if use_dendrites:
+                add_sparse_dendrite_layer(
+                    network=self,
+                    suffix=i + 1,
+                    in_dim=input_size,
+                    out_dim=linear_n[i],
+                    dpc=self.dpc,
+                    threshold=2.,
+                    use_batch_norm=use_batch_norm,
+                    weight_sparsity=linear_weight_percent_on[i],
+                    percent_on=linear_activity_percent_on[i],
+                    k_inference_factor=k_inference_factor,
+                    boost_strength=boost_strength,
+                    boost_strength_factor=boost_strength_factor,
+                    duty_cycle_period=duty_cycle_period,
+                )
+            else:
+                add_sparse_linear_layer(
+                    network=self,
+                    suffix=i + 1,
+                    input_size=input_size,
+                    linear_n=linear_n[i],
+                    dropout=dropout,
+                    use_batch_norm=use_batch_norm,
+                    weight_sparsity=linear_weight_percent_on[i],
+                    percent_on=linear_activity_percent_on[i],
+                    k_inference_factor=k_inference_factor,
+                    boost_strength=boost_strength,
+                    boost_strength_factor=boost_strength_factor,
+                    duty_cycle_period=duty_cycle_period,
+                    consolidated_sparse_weights=consolidated_sparse_weights,
+                )
+            input_size = linear_n[i]
+
+        # add_sparse_linear_layer(
+        #     network=self,
+        #     suffix=i + 2,
+        #     input_size=input_size,
+        #     linear_n=11,
+        #     dropout=dropout,
+        #     use_batch_norm=use_batch_norm,
+        #     weight_sparsity=linear_weight_percent_on[i],
+        #     percent_on=linear_activity_percent_on[i],
+        #     k_inference_factor=k_inference_factor,
+        #     boost_strength=boost_strength,
+        #     boost_strength_factor=boost_strength_factor,
+        #     duty_cycle_period=duty_cycle_period,
+        #     consolidated_sparse_weights=consolidated_sparse_weights,
+        # )
+
+        # Classifier
+        # self.add_module("output", nn.Linear(input_size, num_classes))
+        if use_softmax:
+            self.add_module("softmax", nn.LogSoftmax(dim=1))
+
+
+def add_sparse_dendrite_layer(
```

Comment:
There are classes and methods interleaved in the file, it helps if they are grouped (usually classes go on the top, but not a hard requirement). It would also help organization to not leave functions hanging, either define them as a method in a class or have a specific module where several functions which do similar things are grouped.